mobile/ios/ios-device.ts

```diff
@@ -107,8 +126,126 @@ export class IOSDevice implements Mobile.IIOSDevice {
 	}
 
 	private stopSession() {
-			var result = this.$mobileDevice.deviceStopSession(this.devicePointer);
-			this.validateResult(result, "Unable to stop session");
+		var result = this.$mobileDevice.deviceStopSession(this.devicePointer);
+		this.validateResult(result, "Unable to stop session");
+	}
+
+	private getDeviceValue(value: string): string {
+		var deviceCopyValue = this.$mobileDevice.deviceCopyValue(this.devicePointer, null, this.$coreFoundation.createCFString(value));
+		return this.$coreFoundation.convertCFStringToCString(deviceCopyValue);
+	}
+
+	private lookupApplications(): IDictionary<any> {
+		this.connect();
+		try {
+			this.startSession();
+			try {
+				var dictionaryPointer = ref.alloc(CoreTypes.cfDictionaryRef);
+				var result = this.$mobileDevice.deviceLookupApplications(this.devicePointer, 0, dictionaryPointer);
+				if(result !== 0) {
+					this.$errors.fail("Invalid result code from device lookup applications");
+				}
+				var cfDictionary = dictionaryPointer.deref();
+				var jsDictionary = this.$coreFoundation.cfTypeTo(cfDictionary);
+				return jsDictionary;
+			} finally {
+				this.stopSession();
+			}
+		} finally {
+			this.disconnect();
+		}
+	}
+
+	private findDeveloperDirectory(): IFuture<string> {
+		return (() => {
+			var childProcess = this.$childProcess.spawnFromEvent("xcode-select", ["-print-path"], "close").wait();
+			return childProcess.stdout.trim();
+		}).future<string>()();
+	}
+
+	private findDeveloperDiskImageDirectoryPath(): IFuture<string> {
+		return (() => {
+			var developerDirectory = this.findDeveloperDirectory().wait();
+			var buildVersion = this.getDeviceValue("BuildVersion");
+			var productVersion = this.getDeviceValue("ProductVersion");
+			var productVersionParts = productVersion.split(".");
+			var productMajorVersion = productVersionParts[0];
+			var productMinorVersion = productVersionParts[1];
+
+			var developerDiskImagePath = path.join(developerDirectory, "Platforms", "iPhoneOS.platform", "DeviceSupport");
+			var supportPaths = this.$fs.readDirectory(developerDiskImagePath).wait();
+
+			var supportPath: any = null;
+
+			_.each(supportPaths, (sp: string) => {
+				var parts = sp.split(' ');
+				var version = parts[0];
+				var versionParts = version.split(".");
+
+				var supportPathData = {
+					version: version,
+					majorVersion: versionParts[0],
+					minorVersion: versionParts[1],
+					build: parts.length > 1 ? parts[1].replace("(", "").replace(")", "") : null,
+					path: path.join(developerDiskImagePath, sp)
+				}
+
+				if(supportPathData.majorVersion === productMajorVersion) {
+					if(!supportPath) {
+						supportPath = supportPathData;
+					} else {
+						// is this better than the last match?
+						if(supportPathData.minorVersion === productMinorVersion) {
```

Comment:
Unfortunately I could'n simplify this code. If you have any concrete idea I am open to any suggestion.